extensions/amp-a4a/0.1/legacy-signature-verifier.js

```diff
@@ -14,51 +14,302 @@
+   */
+  loadKeyset(signingServiceName, waitFor) {
+    dev().assert(getMode().localDev || !endsWith(signingServiceName, '-dev'));
+    if (!this.isAvailable_()) {
+      return;
+    }
+    const url = signingServerURLs[signingServiceName];
+    if (url) {
+      // Delay request until document is not in a prerender state.
+      this.keys_.push(waitFor.then(() =>
+          Services.xhrFor(this.win_).fetchJson(url, {
+            mode: 'cors',
+            method: 'GET',
+            // Set ampCors false so that __amp_source_origin is not
+            // included in XHR CORS request allowing for keyset to be cached
+            // across pages.
+            ampCors: false,
+            credentials: 'omit',
+          })).then(res => res.json())
+          .then(jwkSetObj => {
+            if (jwkSetObj && isArray(jwkSetObj['keys']) &&
+                jwkSetObj['keys'].every(isObject)) {
+              return jwkSetObj['keys'];
+            } else {
+              user().error(
+                  TAG,
+                  `Invalid response from signing server ${signingServiceName}`);
+              return [];
+            }
+          }).then(jwks => ({
+            signingServiceName,
+            keys: jwks.map(jwk => this.importPublicKey_(signingServiceName, jwk)
+                .catch(err => {
+                  user().error(
+                      TAG, `error importing keys for: ${signingServiceName}`,
+                      err);
+                  return null;
+                })),
+          })).catch(err => {
+            user().error(TAG, err);
+            // TODO(a4a-team): This is a failure in the initial attempt to get
+            // the keys, probably b/c of a network condition.  We should
+            // re-trigger key fetching later.
+            return {signingServiceName, keys: []};
+          }));
+    } else {
+      // The given signingServiceName does not have a corresponding URL in
+      // _a4a-config.js.
+      const reason = `Signing service '${signingServiceName}' does not exist.`;
+      user().error(TAG, reason);
+      this.keys_.push(/** @type {!Promise<!CryptoKeysDef>} */ (Promise.resolve({
+        signingServiceName,
+        keys: [],
+      })));
+    }
+  }
+
+  /**
+   * Extracts a cryptographic signature from `headers` and verifies that it's
+   * the correct cryptographic signature for `creative`.
+   *
+   * As a precondition, `loadKeyset` must have already been called on the
+   * signing service that was used.
+   *
+   * @param {!ArrayBuffer} creative
+   * @param {!Headers} headers
+   * @param {function(string, !Object)} lifecycleCallback called for each AMP
+   *     lifecycle event triggered during verification
+   * @return {!Promise<?VerificationFailure>} resolves to `null` on success, or,
+   *     on failure, to a `VerificationFailure` indicating the cause
+   */
+  verify(creative, headers, lifecycleCallback) {
+    if (!this.isAvailable_()) {
+      return Promise.resolve(
```

Comment:
Those `error` calls were copied over from this code's previous home in `amp-a4a.js`. At some point they need to be cleaned up.

Do `user().error` logs go to Stackdriver, or is this just for the Dev Tools? The Dev Tools will already log a failed network request. I'm not sure it makes sense to log connectivity failures to StackDriver. If we're worried about CORS, maybe it would make sense to downsample them or something?